compiler/noirc_frontend/src/tests.rs

```diff
@@ -1444,3 +1451,328 @@ fn specify_method_types_with_turbofish() {
+    }
+    
+    trait Deserialize<let N: u64> {
+        fn deserialize(fields: [Field; N]) -> Self;
+    }
+    "#;
+    let errors = get_program_errors_elaborator(src);
+    assert!(errors.is_empty());
+}
+
+#[test]
+fn numeric_generic_used_in_where_clause() {
+    let src = r#"
+
+    trait Deserialize<let N: u32> {
+        fn deserialize(fields: [Field; N]) -> Self;
+    }
+
+    fn read<T, let N: u32>() -> T where T: Deserialize<N> {
+        let mut fields: [Field; N] = [0; N];
+        for i in 0..N {
+            fields[i] = i as Field + 1;
+        }
+        T::deserialize(fields)
+    }
+    "#;
+    let errors = get_program_errors_elaborator(src);
+    assert!(errors.is_empty());
+}
+
+#[test]
+fn implicit_numeric_generics_elaborator() {
+    let src = r#"
+    struct BoundedVec<T, MaxLen> {
+        storage: [T; MaxLen],
+        len: u64,
+    }
+    
+    impl<T, MaxLen> BoundedVec<T, MaxLen> {
+
+        // Test that we have an implicit numeric generic for "Len" as well as "MaxLen"
+        pub fn extend_from_bounded_vec<Len>(&mut self, _vec: BoundedVec<T, Len>) { 
+            // We do this to avoid an unused variable warning on `self`
+            let _ = self.len;
+            for _ in 0..Len { }
+        }
+
+        pub fn push(&mut self, elem: T) {
+            assert(self.len < MaxLen, "push out of bounds");
+            self.storage[self.len] = elem;
+            self.len += 1;
+        }
+    }
+    "#;
+    let errors = get_program_errors_elaborator(src);
+
+    for error in errors.iter() {
+        if let CompilationError::ResolverError(ResolverError::UseExplicitNumericGeneric { ident }) =
+            &errors[0].0
+        {
+            assert!(matches!(ident.0.contents.as_str(), "MaxLen" | "Len"));
+        } else {
+            panic!("Expected ResolverError::UseExplicitNumericGeneric but got {:?}", error);
+        }
+    }
+}
+
+#[test]
+fn nested_generic_elaborator() {
+    let src = r#"
+    trait Default {
+        fn default() -> Self;
+    }
+
+    struct Option<T> {
+        _is_some: bool,
+        _value: T,
+    }
+
+    impl<T> Option<T> {
+        pub fn flatten(option: Option<Option<T>>) -> Option<T> where T: Default {
+            if option._is_some {
+                option._value
+            } else {
+                Self { _is_some: false, _value: T::default() }
+            }
+        }
+    }
+    "#;
+    let errors = get_program_errors_elaborator(src);
+    assert!(errors.is_empty());
+}
+
+#[test]
+fn single_parser_error_bad_numeric_generic() {
+    let src = r#"
+    fn id<let I>(x: [Field; I]) -> [Field; I] {
+        x
+    }
+    "#;
+    let errors = get_program_errors_elaborator(src);
+    assert!(has_parser_error(&errors));
+    // We want to make sure that each ill-formed numeric generic (e.g. `<let I>` or `<I:>`)
+    // returns a single parser error.
+    // We have this test to make sure we are not displaying deceiving errors to the user.
+    assert_eq!(errors.len(), 1);
+}
```

Comment:
Shouldn't this be a parser test instead?

Reply:
I'll move it to the parser